### DISCUSSION TITLE
How to make a HTTP signature for mastodon in python successsfully?

### DISCUSSION THREAD

**Discussion by Yoxem** [mastodon/mastodon#18600] [Q&A] [answered]
I have followed the tutorial https://github.com/timmot/activity-pub-tutorial (under MIT License) to try to make a connection to Mastodon.

However, when I tried to make a following request with a HTTP signature to a mastodon instance, the instance always return 401 Error. it means that the verification failed.

Because it's complex to make a sign with sha256 and then convert it to base64, I think that it must be something wrong with the signing process. How to solve the problem and how to find what is wrong.

The request-sending script (Python) with modification is shown below.

The request send to Pleoroma successfully (returning 200), while sending to Mastodon return 401 (without related cache issue).

```
from cryptography.hazmat.backends import default_backend as crypto_default_backend
from cryptography.hazmat.primitives import serialization as crypto_serialization
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding

from urllib.parse import urlparse
import base64
import datetime
import requests


recipient_url = "https://orig.example.social/users/alice"
recipient_inbox = "https://orig.example.social/users/alice/inbox"

sender_url = "https://dest.example.net/users/bob"
sender_key = "https://dest.example.net/users/bob#main-key"

activity_id = "https://dest.example.net/users/bob/follows/test"


# The following is to sign the HTTP request as defined in HTTP Signatures.
private_key_text = open('./lianlok/private.pem', 'r').read().encode('ascii') # load from file

private_key = crypto_serialization.load_pem_private_key(
    private_key_text,
    password=None,
    backend=crypto_default_backend()
)

current_date = datetime.datetime.utcnow().strftime('%a, %d %b %Y %H:%M:%S GMT')

recipient_parsed = urlparse(recipient_inbox)
recipient_host = recipient_parsed.netloc
recipient_path = recipient_parsed.path

signature_text = b'(request-target): post %s\nhost: %s\ndate: %s' % (recipient_path.encode('ascii'), recipient_host.encode('ascii'), current_date.encode('ascii'))

raw_signature = private_key.sign(
    signature_text,
    padding.PKCS1v15(),
    hashes.SHA256()
)

signature_header = 'keyId="%s",headers="(request-target) host date",signature="%s"' % (sender_key, base64.b64encode(raw_signature).decode('ascii'))

headers = {
    'Date': current_date,
    'Content-Type': 'application/activity+json',
    'Host': recipient_host,
    'Signature': signature_header
}

# Now that the header is set up, we will construct the message
follow_request_message = {
    "@context": "https://www.w3.org/ns/activitystreams",
    "id": activity_id,
    "type": "Follow",
    "actor": sender_url,
    "object": recipient_url
}

r = requests.post(recipient_inbox, headers=headers, json=follow_request_message)

print(r)
```

**Comment by ClearlyClaire** [comment, marked as answer]
I took the liberty of editing the topic and post to change occurrences of “HTML signatures” to “HTTP signatures”.

Something useful to know when debugging this kind of issues is that you can sometime get some more information from Mastodon by reading the reply's body, even on 401 (though it may also return the somewhat unhelpful `Public key not found for key <URL>` which indicates an issue in fetching the actor's key itself; or `Verification failed for <acct> <uri> using rsa-sha256 (RSASSA-PKCS1-v1_5 with SHA-256)` that indicates an issue with the signature itself).

In this particular case, it seems you are not providing a `Digest` header, that Mastodon requires for `POST` requests.

**Reply by Yoxem** [reply]
Thank you very much. After adding digest, it works., the result code is like: https://github.com/timmot/activity-pub-tutorial/issues/1#issuecomment-1147449461

**Comment by Yoxem** [comment]
I'm also curious about that: how to verify a "Note" json request with a digest or a signature?
I tried to convert from the json to a sha256 hash value, but it doesn't matched to the digest value in the header.

**Reply by nightpool** [reply]
@Yoxem verifying an HTTP signature (Signature header) is exactly the same as constructing the signature string, just in reverse. If you have working code for constructing a Signature header, you should use the exact same code for verifying the Signature header.

**Reply by ClearlyClaire** [reply]
Compute the digest of the response's body prior to parsing the JSON, as encoding the already-parsed JSON may cause a different output.
